Feature/PROD-1292 | Update api docs to reflect new msdp search functionality

sunny-trip-ninja opened the pull request:
## LocalQA approved by
@Sanpele 

## Description
- Updated the MSDP search documentation including
  - Search
  - Get Flight results
  - Get Hotel results
  - Remove hotel results 

## How to Test
1. Checkout this branch
2. Run  `npx @redocly/openapi-cli preview-docs tn_api_msdp_v1.yml`
3. Go to http://127.0.0.1:8080
4. You should see the documentation for the mentioned endpoints above in `Description`

mschembri-TN commented:
Task linked: [PROD-1292 Update API Docs to reflect new MSDP Search functionality](https://app.clickup.com/t/14197839/PROD-1292)

Sanpele commented on tn_msdp_api_v1.yml:
```diff
@@ -0,0 +1,1736 @@
+                    application/json:
+                        schema:
+                            $ref: "#/components/schemas/Requests/MSDPSearchRequest"
+            responses:
+                200:
+                    description: Success
+                    content:
+                        application/json:
+                            schema:
+                                $ref: "#/components/schemas/Responses/MSDPSearchResponse"
```
Might be useful to return OTA and QT format of response side by side instead of altogether. I am looking todo the same thing in https://app.clickup.com/t/14197839/ENG-672. 

Here is a source i used to do this : https://stackoverflow.com/a/48427558
Look at the OpenAPI 3.x section.

This is what it looks like to have more then 1 response on my branch. 

![Screenshot 2022-12-05 at 4 08 52 PM](https://user-images.githubusercontent.com/47077094/205752951-56755f4f-5f42-465a-a213-0a604e025a0e.png)

![Screenshot 2022-12-05 at 4 07 38 PM](https://user-images.githubusercontent.com/47077094/205752774-22d1e331-ab4a-40f3-8193-c9c822890482.png)

peterdrohan replied:
this is pretty slick !

sunny-trip-ninja replied:
ooh boy, thats very advanced and it looks amazing, will take a look when the the higher priority tickets are finished

sunny-trip-ninja replied:
@Sanpele Updated!

flutistar commented on tn_msdp_api_v1.yml:
```diff
@@ -0,0 +1,1736 @@
+    title: Trip Ninja API Documentation
+    description: |
+        <h2>Introduction</h2>
+        Trip Ninja uses a REST API with requests and responses in JSON format. For development and testring, the pre-production servers
```
```suggestion
        Trip Ninja uses a REST API with requests and responses in JSON format. For development and testing, the pre-production servers
```

flutistar commented on tn_msdp_api_v1.yml:
```diff
@@ -0,0 +1,1736 @@
+        and pre-production endpoints - these correspond to the GDS’s production and pre-production services, and can be used as desired.<br  />
+        <h2>Trip Ninja API URLs</h2>
+        
+        Pre-Production: https://preprodapi.tripninja.io
```
One question:
The url will looks like this?
`https://preprodapi.tripninja.io/preprod/` or `https://preprodapi.tripninja.io/prod/`?

Suggestion:
URL: `https://preprod.api.tripninja.io/` 
`{endpoint}`: I think we should be able to get `endpoint` variable on the backend side.

sunny-trip-ninja replied:
in the docs the endpoints will look like this, so i think the current setup is alright 
![image](https://user-images.githubusercontent.com/99681496/205775599-ed7cab94-c4d0-414f-916a-85a79f83b39d.png)

flutistar replied:
I don't mean the document itself.
Let's assume that we are API users.
To send the request to our API, the url will looks like this - `https://preprodapi.tripninja.io/preprod/`.
Here we have double `preprod` on the url and I think it should be updated.
What do you think?

sunny-trip-ninja replied:
https://app.clickup.com/t/14197839/PROD-1735

sunny-trip-ninja replied:
done! it only shows prod now! @flutistar

sunny-trip-ninja commented on tn_msdp_api_v1.yml:
```diff
@@ -0,0 +1,1736 @@
+                                    description: Parameter toggles which travelport regional service endpoint you hit *Travelport Only*
+                                    type: string
+                                    example: "americas"
+                                itx:
```
plz remove this

Sanpele commented:
localQA in progress

Sanpele commented:
few smal things that may be good to tweak. Also had a few bigger ideas but those should prob be separate stuffs. 

SMALLER THINGS:
1. The credentials in the example are Amadeus, but its impossible to run a MSDP search with just Amadeus credentials, maybe change to TP or add a WebBeds credential aswell?
2. In the response samples, for the include_itineraries = false ones the flight results have an extra list around segments.
![Screenshot 2022-12-22 at 10 50 52 AM](https://user-images.githubusercontent.com/47077094/209186283-70e0528e-619e-454d-b9fa-67147bf936e2.png)
![Screenshot 2022-12-22 at 10 51 35 AM](https://user-images.githubusercontent.com/47077094/209186339-a4962015-e1ce-4f24-b233-6e3ec88246f6.png)
3. in the response description for the search response, the msdp_search_details dosen't have a description. 
![Screenshot 2022-12-22 at 10 52 18 AM](https://user-images.githubusercontent.com/47077094/209186441-829680a9-fde9-4a15-b42d-0c8435678037.png)
4. bit more detail here might be good, like is it all results or 1 flight. What it may be used for. Not sure tho, wat u think?
![Screenshot 2022-12-22 at 10 57 53 AM](https://user-images.githubusercontent.com/47077094/209187329-91dfab6a-3212-4ef5-93e1-fdd89bb62e85.png)
5. Bit more detail here would be good as-well i think, like specifying its only for 1 hotel search, and that you can change the order by all the fields if you want. Also having some space between the example would be good i think. We so have an example URL already, but i agree because this one is detailed it makes sense to double up. 
![Screenshot 2022-12-22 at 10 58 58 AM](https://user-images.githubusercontent.com/47077094/209187700-dfbe55a6-9976-4ed8-8b96-2d80b3d32a4a.png)
![Screenshot 2022-12-22 at 10 59 03 AM](https://user-images.githubusercontent.com/47077094/209187714-e57cc52b-1128-48f0-9a14-6035f02f6a47.png)
6. the msdp_package_id, trip_id and hotel_trace_id's in the response should be different from each other, and of the normal length we return to match what we will return when users actually use the msdp search. This spans all the endpoints in the docs.
![Screenshot 2022-12-22 at 11 04 40 AM](https://user-images.githubusercontent.com/47077094/209188300-0748fee2-0fb1-4ad8-b103-94c682f62dd2.png)
![Screenshot 2022-12-22 at 11 04 47 AM](https://user-images.githubusercontent.com/47077094/209188305-82a942d9-592d-4408-976c-3c507ab1c897.png)
![Screenshot 2022-12-22 at 11 04 54 AM](https://user-images.githubusercontent.com/47077094/209188315-1d95c830-adb6-4414-80c5-f5cc047ba846.png)



BIGGER THINGS DEF OUTSIDE SCOPE:
7. We should include a section at the top of the doc that explains/sells MSDP/PACKAGES. Sort of like what we have for farestructure now. This seems like something product could help with / do. wat u think?  
8. We should have better navigation between Flight/Hotels/Pricing/MSDP now. The only way i know how it to modify the url and that deff not ideal. Not sure if ther is a way i don't know about. maybe adding some way to navigate here. wat u think, ticket material?
![Screenshot 2022-12-22 at 10 46 00 AM](https://user-images.githubusercontent.com/47077094/209184908-f9555fe2-870e-4612-9aa0-e75562fb22c0.png)
9. We should include more info about the validation required in the MSDP search section for sure, maybe the modify sections as-well. Will save users some time. @peterdrohan maybe could write/add a summary in a separate ticket? 
10. The response codes are a little out of date, there are a bunch of new MSDP ones but i think in general they are out of date. It may be worth having a ticket where someone cleans them up a bit on the API side, then after that someone could update the docs with them per endpoint. As right now we only really include the basic 200/206/400/401 but have many more options. wat u think?



QUESTIONS:
11. The version RN is 1.0.0 but the title is Generic Trip ninja API. Are we having a seperate MSDP version? if yea we should update the title, otherwise should update the version to our current doc one.

sunny-trip-ninja commented:
@Sanpele Tysm for doing the QA, it was really detailed
Addressed comments 1-6, 11

7) Yeah we should ask Product for a brief description for MSDP
8, 9 , 10) Lets create separate tickets and have it in the backlog

Sanpele commented:
Have created some tickets, will re-QA now:

8 - improve navigation between sections of our docs.
https://app.clickup.com/t/14197839/ENG-1076
9 - add validation details to MSDP search docs.
https://app.clickup.com/t/14197839/ENG-1077
10 - Have discussion on our current usuage of error codes in API. 
https://app.clickup.com/t/14197839/ENG-1078

sunny-trip-ninja commented:
@Sanpele also updated 7) !

Sanpele commented:
Found couple more small things i missed in initial QA. Been clickin into things like mad but once these fixed ill happily approve looking real good! 

1. The address in hotel results dosen't have a description:
![Screenshot 2022-12-23 at 10 24 16 AM](https://user-images.githubusercontent.com/47077094/209367406-a2954ed9-83af-49c9-9bb9-912f215fed57.png)
2. The credential_info in the response still is displaying required fields, may be worth separating a request/response credential_info component. 
![Screenshot 2022-12-23 at 10 11 47 AM](https://user-images.githubusercontent.com/47077094/209367600-d6983b21-044b-42d7-a4e6-cf68f52fb8e2.png)
3. Flight array for a include_itineraries = false search response has a few fields missing description.
![Screenshot 2022-12-23 at 10 28 06 AM](https://user-images.githubusercontent.com/47077094/209367969-98036be8-3f21-448a-bb2e-f0cfc17088c3.png)
![Screenshot 2022-12-23 at 10 28 51 AM](https://user-images.githubusercontent.com/47077094/209367975-33bc9431-c3a2-4540-8b2a-3c1dfbb40899.png)
4. Extra lower case a here
![Screenshot 2022-12-23 at 10 19 34 AM](https://user-images.githubusercontent.com/47077094/209368043-eaa4ad61-7991-4f2a-9581-690b40d5b4ed.png)
5. Was thinking, would good to mention here that a MSDP search much have at least 1 hotel present, and an error will be thrown if you attempt to remove the last hotel in a MSDP package. 
![Screenshot 2022-12-23 at 10 30 45 AM](https://user-images.githubusercontent.com/47077094/209368290-6a8c8bef-593b-47ba-9441-2883d469c15b.png)
Something like: 
This endpoint removes a specific hotel result from the prior msdp search. All msdp packages must have at least 1 hotel. An error will be thrown if you attempt to remove the final hotel from a package.

sunny-trip-ninja commented:
> Found couple more small things i missed in initial QA. Been clickin into things like mad but once these fixed ill happily approve looking real good!
> 
> 1. The address in hotel results dosen't have a description:
>    ![Screenshot 2022-12-23 at 10 24 16 AM](https://user-images.githubusercontent.com/47077094/209367406-a2954ed9-83af-49c9-9bb9-912f215fed57.png)
> 2. The credential_info in the response still is displaying required fields, may be worth separating a request/response credential_info component.
>    ![Screenshot 2022-12-23 at 10 11 47 AM](https://user-images.githubusercontent.com/47077094/209367600-d6983b21-044b-42d7-a4e6-cf68f52fb8e2.png)
> 3. Flight array for a include_itineraries = false search response has a few fields missing description.
>    ![Screenshot 2022-12-23 at 10 28 06 AM](https://user-images.githubusercontent.com/47077094/209367969-98036be8-3f21-448a-bb2e-f0cfc17088c3.png)
>    ![Screenshot 2022-12-23 at 10 28 51 AM](https://user-images.githubusercontent.com/47077094/209367975-33bc9431-c3a2-4540-8b2a-3c1dfbb40899.png)
> 4. Extra lower case a here
>    ![Screenshot 2022-12-23 at 10 19 34 AM](https://user-images.githubusercontent.com/47077094/209368043-eaa4ad61-7991-4f2a-9581-690b40d5b4ed.png)
> 5. Was thinking, would good to mention here that a MSDP search much have at least 1 hotel present, and an error will be thrown if you attempt to remove the last hotel in a MSDP package.
>    ![Screenshot 2022-12-23 at 10 30 45 AM](https://user-images.githubusercontent.com/47077094/209368290-6a8c8bef-593b-47ba-9441-2883d469c15b.png)
>    Something like:
>    This endpoint removes a specific hotel result from the prior msdp search. All msdp packages must have at least 1 hotel. An error will be thrown if you attempt to remove the final hotel from a package.

All Addressed, thanks for doing the QA!

Sanpele commented:
@sunny-trip-ninja localQA approved! docs look great and new changes are up and running!
